raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java

```diff
@@ -103,6 +111,58 @@ public void testInitializeSingleMemberQuorum() throws IOException {
         assertEquals(0, channel.drainSendQueue().size());
     }
 
+    @Test
+    public void testInitializeAsLeaderFromStateStore() throws IOException {
+        Set<Integer> voters = Utils.mkSet(localId, 1);
+        MockQuorumStateStore electionStore = new MockQuorumStateStore() {
+            @Override
+            public ElectionState readElectionState() {
+                return ElectionState.withElectedLeader(2, localId);
+            }
+        };
+        KafkaRaftClient client = buildClient(voters, electionStore);
+        assertTrue(stateMachine.state().isPresent());
+        assertTrue(stateMachine.state().get() instanceof LeaderState);
+        assertTrue(client.quorumState().isLeader());
+        assertEquals(1L, log.endOffset());
+
+        // Should have sent out connection info query for other node id.
+        client.poll();
+        List<RaftMessage> unsent = channel.drainSendQueue();
+        assertEquals(1, unsent.size());
+
+        assertTrue(unsent.get(0) instanceof RaftRequest.Outbound);
+        assertTrue(unsent.get(0).data() instanceof FindQuorumRequestData);
+
+        assertTrue(unsent.get(0) instanceof RaftRequest.Outbound);
+        assertEquals(-1, ((RaftRequest.Outbound) unsent.get(0)).destinationId());
+    }
+
+    @Test
+    public void testInitializeAsCandidateFromStateStore() throws IOException {
+        // Need 3 node to require a 2-node majority
+        Set<Integer> voters = Utils.mkSet(localId, 1, 2);
+        MockQuorumStateStore electionStore = new MockQuorumStateStore() {
+            @Override
+            public ElectionState readElectionState() {
+                return ElectionState.withVotedCandidate(2, localId);
+            }
+        };
+        KafkaRaftClient client = buildClient(voters, electionStore);
+        assertFalse(stateMachine.state().isPresent());
+        assertEquals(0L, log.endOffset());
+        assertTrue(client.quorumState().isCandidate());
+
+        // Should have sent out connection info query for other node id.
+        client.poll();
+        List<RaftMessage> unsent = channel.drainSendQueue();
+        assertEquals(1, unsent.size());
+
+        assertTrue(unsent.get(0) instanceof RaftRequest.Outbound);
```

Comment:
For this test case to be more useful, I think we should add assertions on the expected vote requests as a candidate.

Reply:
Are you suggesting to receive a mock vote request?

Reply:
NVM, I think I understood.